velox/substrait/VeloxToSubstraitExpr.cpp

```diff
@@ -0,0 +1,383 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/substrait/VeloxToSubstraitExpr.h"
+
+#include "velox/expression/ControlExpr.h"
+
+namespace facebook::velox::substrait {
+
+void VeloxToSubstraitExprConvertor::toSubstraitExpr(
+    const std::shared_ptr<const ITypedExpr>& vExpr,
+    RowTypePtr vPreNodeOutPut,
+    ::substrait::Expression* sExpr) {
+  if (std::shared_ptr<const ConstantTypedExpr> vConstantExpr =
+          std::dynamic_pointer_cast<const ConstantTypedExpr>(vExpr)) {
+    toSubstraitLiteral(vConstantExpr->value(), sExpr->mutable_literal());
+    return;
+  }
+  if (auto vCallTypeExpr =
+          std::dynamic_pointer_cast<const CallTypedExpr>(vExpr)) {
+    std::shared_ptr<const Type> vExprType = vCallTypeExpr->type();
+    std::vector<std::shared_ptr<const ITypedExpr>> vCallTypeInputs =
+        vCallTypeExpr->inputs();
+    std::string vCallTypeExprFunName = vCallTypeExpr->name();
+    // different by function names.
+    if (vCallTypeExprFunName == exec::kIf) {
+      ::substrait::Expression_IfThen* sFun = sExpr->mutable_if_then();
+      int64_t vCallTypeInputSize = vCallTypeInputs.size();
+      for (int64_t i = 0; i < vCallTypeInputSize; i++) {
+        std::shared_ptr<const ITypedExpr> vCallTypeInput =
+            vCallTypeInputs.at(i);
+        // TODO
+        //  need to judge according the names in the expr, and then set them to
+        //  the if or then or else expr can debug to find when process project
+        //  node
+      }
+    } else if (vCallTypeExprFunName == exec::kSwitch) {
+      ::substrait::Expression_SwitchExpression* sFun =
+          sExpr->mutable_switch_expression();
+      // TODO
+    } else {
+      ::substrait::Expression_ScalarFunction* sFun =
+          sExpr->mutable_scalar_function();
+      // TODO need to change yaml file to register functin, now is dummy.
+      // the substrait communcity have changed many in this part...
+      uint32_t sFunId =
+          v2SFuncConvertor_.registerSubstraitFunction(vCallTypeExprFunName);
+      sFun->set_function_reference(sFunId);
+
+      for (auto& vArg : vCallTypeInputs) {
+        ::substrait::Expression* sArg = sFun->add_args();
+        toSubstraitExpr(vArg, vPreNodeOutPut, sArg);
+      }
+      ::substrait::Type* sFunType = sFun->mutable_output_type();
+      v2STypeConvertor_.toSubstraitType(vExprType, sFunType);
+      return;
+    }
+  }
+  if (auto vFieldExpr =
```

Comment:
same here. It would be nice to be consistent and have one `toSubstraitExpr()` overload handling each type of expression.

Reply:
done.